fairlearn/postprocessing/_cvxpy_threshold_optimizer.py

```diff
@@ -0,0 +1,395 @@
+        of groups).
+
+        Returns
+        -------
+        float
+            The equal-odds constraint violation.
+        """
+        self._check_fit_status()
+
+        n_groups = len(self.groupwise_roc_points)
+
+        # Compute l-inf distance between each pair of groups
+        linf_constraint_violation = [
+            (np.linalg.norm(
+                self.groupwise_roc_points[i] - self.groupwise_roc_points[j],
+                ord=np.inf), (i, j))
+            for i, j in product(range(n_groups), range(n_groups))
+            if i < j
+        ]
+
+        # Return the maximum
+        max_violation, (groupA, groupB) = max(linf_constraint_violation)
+        logging.info(
+            f"Maximum fairness violation is between "
+            f"group={groupA} (p={self.groupwise_roc_points[groupA]}) and "
+            f"group={groupB} (p={self.groupwise_roc_points[groupB]});"
+        )
+
+        return max_violation
+
+
+    def fit(self, X: np.ndarray, y: np.ndarray, group: np.ndarray, y_scores: np.ndarray = None):
+        """Fit this predictor to achieve the (possibly relaxed) equal odds 
+        constraint on the provided data.
+
+        Parameters
+        ----------
+        X : np.ndarray
+            The input features.
+        y : np.ndarray
+            The input labels.
+        group : np.ndarray
+            The group membership of each sample.
+            Assumes groups are numbered [0, 1, ..., num_groups-1].
+        y_scores : np.ndarray, optional
+            The pre-computed model predictions on this data.
+
+        Returns
+        -------
+        callable
+            Returns self.
+        """
+
+        # Compute group stats
+        self._global_prevalence = np.sum(y) / len(y)
+
+        unique_groups = np.unique(group)
+        num_groups = len(unique_groups)
+        if np.max(unique_groups) > num_groups-1:
+            raise ValueError(
+                f"Groups should be numbered starting at 0, and up to "
+                f"num_groups-1. Got {num_groups} groups, but max value is "
+                f"{np.max(unique_groups)} != num_groups-1 == {num_groups-1}."
+            )
+
+        # Relative group sizes for LN and LP samples
+        group_sizes_label_neg = np.array([
+            np.sum(1 - y[group == g]) for g in unique_groups
+        ])
+        group_sizes_label_pos = np.array([
+            np.sum(y[group == g]) for g in unique_groups
+        ])
+
+        if np.sum(group_sizes_label_neg) + np.sum(group_sizes_label_pos) != len(y):
+            raise RuntimeError(f"Failed sanity check. Are you using non-binary labels?")
+
+        # Convert to relative sizes
+        group_sizes_label_neg = group_sizes_label_neg.astype(float) / np.sum(group_sizes_label_neg)
+        group_sizes_label_pos = group_sizes_label_pos.astype(float) / np.sum(group_sizes_label_pos)
+
+        # Compute group-wise ROC curves
+        if y_scores is None:
+            y_scores = self.estimator(X)
```

Comment:
We're doing this a bit differently in TO and I would argue it's preferable: https://github.com/fairlearn/fairlearn/blob/476db8c7ca648a4f9f078e91bc3b4058c3a5558a/fairlearn/postprocessing/_threshold_optimizer.py#L334

Reply:
ah right, that was still using the previous callable API.

Reply:
By the way, we can of course omit the `y_scores` here, it's just here because this is the slowest part of the whole `fit` method and oftentimes users already have the predictions computed.

An example is when you want to map the whole fairness-accuracy Pareto frontier, you'd call `RelaxedThresholdOptimizer` on a bunch of tolerance values (e.g., `np.arange(0, 1, 1e-2)`), and each call would unnecessarily re-compute predictions. This is in fact the only part of the fit method that scales with the size of the dataset, everything else is generally fast on large datasets.

Reply:
As an aside, I noticed that the default `predict_method` for `ThresholdOptimizer` is `predict`, when for postprocessing algorithms it should really be `predict_proba`, as we want a risk score instead of binary predictions.

Should this be changed in the current implementation?

Reply:
I get your point about performance, but I'm not entirely sure why a user would precompute them just to pass them in manually, rather than having the class compute the scores. Or do you mean that they need the scores for other purposes, so they *have to* compute them outside, so computing them twice feels like a waste?

Reply:
hmm, I don't have a strong opinion on how exactly it is implemented, but I think it would be useful to somehow cache these predictions so we don't have to re-compute them for different levels of tolerance. Given that different values of tolerance will be handled by different class objects, I don't see how it could be cached at the class level (must be cached outside). Example follows:

When mapping the fairness-accuracy Pareto frontier, the user needs to create a different `RelaxedThresholdOptimizer` for each `tolerance` value, e.g.:
```py
# Compute predictions for models with varying levels of tolerance
def compute_test_predictions_with_relaxed_constraints(tolerance: float, y_scores=None) -> np.ndarray:
    # Instantiate
    clf = _RelaxedThresholdOptimizer(
        predictor=lambda *args, **kwargs: unmitigated_predictor.predict(*args, **kwargs),
        predict_method="__call__",
        constraint="equalized_odds",
        tolerance=tolerance,
    )

    # NOTE
    # 1st Option: will call `predictor.predict(X_train)`
    clf.fit(X_train, Y_train, sensitive_features=A_train_np)

    # 2nd Option: will *not* call `predictor.predict(X_train)`
    clf.fit(X_train, Y_train, sensitive_features=A_train_np, y_scores=y_scores)

    return clf.predict(X_test, sensitive_features=A_test_np)

# [For 2nd Option] Pre-compute predictions
Y_train_scores = unmitigated_predictor.predict(X_train)

# Compute predictions at different levels of tolerance
all_model_predictions = {
    f"train tolerance={tol:.1}": compute_test_predictions_with_relaxed_constraints(tol, y_scores=Y_train_scores)
    for tol in np.arange(0, unmitigated_equalized_odds_diff, 1e-2)
}
```

Given that the underlying predictor is the same on all of these objects, the `clf.fit(.)` will call `predictor.predict_proba(.)` repeatedly (and redundantly) for each tolerance value. The 1st option calls the predictor's predict_method $n$ times (once per call of the helper function), while the 2nd option only calls it once (outside the function).

PS: ignore the `predict_method="__call__"` mess for now 😅